gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java

```diff
@@ -54,6 +78,87 @@ public class GoogleCloudStorageClientImpl extends ForwardingGoogleCloudStorage {
             .setDownscopedAccessTokenFn(downscopedAccessTokenFn)
             .build());
+    this.storageOptions = checkNotNull(options, "options must not be null");
+    this.storage =
+        clientLibraryStorage == null ? createStorage(credentials, options) : clientLibraryStorage;
+  }
+
+  @Override
+  public WritableByteChannel create(StorageResourceId resourceId, CreateObjectOptions options)
+      throws IOException {
+    if (storageOptions.isGrpcEnabled()) {
+      logger.atFiner().log("create(%s)", resourceId);
+      checkArgument(
+          resourceId.isStorageObject(), "Expected full StorageObject id, got %s", resourceId);
+      // Update resourceId if generationId is missing
+      StorageResourceId resourceIdWithGeneration = resourceId;
+      if (!resourceId.hasGenerationId()) {
+        resourceIdWithGeneration =
+            new StorageResourceId(
+                resourceId.getBucketName(),
+                resourceId.getObjectName(),
+                getWriteGeneration(resourceId, options.isOverwriteExisting()));
+      }
+
+      GoogleCloudStorageClientLibraryWriteChannel channel =
+          new GoogleCloudStorageClientLibraryWriteChannel(
+              storage,
+              storageOptions,
+              resourceIdWithGeneration,
+              options,
+              backgroundTasksThreadPool);
+      channel.initialize();
+      return channel;
+    }
+    return super.create(resourceId, options);
+  }
+
+  @Override
+  public void close() {
+    try {
+      try {
+        super.close();
+      } finally {
+        backgroundTasksThreadPool.shutdown();
+      }
+    } finally {
+      backgroundTasksThreadPool = null;
+    }
+  }
+
+  /**
+   * Gets the object generation for a write operation
+   *
+   * <p>making getItemInfo call even if overwrite is disabled to fail fast in case file is existing.
+   *
+   * @param resourceId object for which generation info is requested
+   * @param overwrite whether existing object should be overwritten
+   * @return the generation of the object
+   * @throws IOException if the object already exists and cannot be overwritten
+   */
+  private long getWriteGeneration(StorageResourceId resourceId, boolean overwrite)
+      throws IOException {
+    logger.atFiner().log("getWriteGeneration(%s, %s)", resourceId, overwrite);
+    GoogleCloudStorageItemInfo info = getItemInfo(resourceId);
+    if (!info.exists()) {
+      return 0L;
+    }
+    if (info.exists() && overwrite) {
+      long generation = info.getContentGeneration();
+      checkState(generation != 0, "Generation should not be 0 for an existing item");
+      return generation;
+    }
+    throw new FileAlreadyExistsException(String.format("Object %s already exists.", resourceId));
+  }
+
+  private static Storage createStorage(
+      Credentials credentials, GoogleCloudStorageOptions storageOptions) {
+    checkNotNull(storageOptions, "options must not be null");
```

Comment:
Seems like this is redundant check, as options are always non-null?

Reply:
removed.